File: wrappers/nodejs/test/did/didweb.test.ts
import { DIDWeb, InMemKms, KeyHandle, KeyType, VerificationMethodKey, VerificationRelationshipType } from "../../";

describe("did:web: ", () => {
  it("generate did document", async () => {
    const did = "did:web:test.example.com";
    const kms = new InMemKms();
    const key = await kms.create(KeyType.P256);
    const nativeKeyHandle = await kms.get(key);
    const didWeb = new DIDWeb();
    let keyHandle: KeyHandle = {
      alg: nativeKeyHandle.alg,
      jwk: nativeKeyHandle.jwk,
      pubKey: nativeKeyHandle.pubKey,
      sign: nativeKeyHandle.sign,
      verify: nativeKeyHandle.verify,
    };
    const verificationMethodKey = new VerificationMethodKey(keyHandle, [
      VerificationRelationshipType.KeyAgreement,
      VerificationRelationshipType.Authentication,
    ]);

    let expected_did_doc = {
      "@context": ["https://www.w3.org/ns/did/v1", "https://w3id.org/security#EcdsaSecp256r1VerificationKey2019"],
      id: did,
      authentication: ["did:web:test.example.com#key-0"],
      keyAgreement: ["did:web:test.example.com#key-0"],
      verificationMethod: [
        {
          id: "did:web:test.example.com#key-0",
          type: "EcdsaSecp256r1VerificationKey2019",
          controller: did,
          publicKeyMultibase: "temp_string",
        },
      ],
    };

    let did_doc = didWeb.generateDidDocument(did, [verificationMethodKey]);
    expected_did_doc.verificationMethod[0].publicKeyMultibase = did_doc.verificationMethod[0][
      "publicKeyMultibase"
    ] as string;
    expect(did_doc).toEqual(expected_did_doc);
  });
});

File: wrappers/uniffi/kotlin/src/test/kotlin/com/bci/DidWebTest.kt
package com.bci

import com.bci.asdk.*
import junit.framework.TestCase.assertEquals
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.BeforeAll
import kotlin.test.Test

class DidWebTest {
    companion object {
        @JvmStatic
        @BeforeAll
        fun setup() {
            setJniLibPath()
        }
    }

    @Test
    fun generateDidDocument() = runTest {
        val did = "did:web:test.example.com"
        val kms = InMemKms()
        val kid = kms.create(KeyType.P256)
        val keyHandle = kms.get(kid)

        val verificationMethodKey = VerificationMethodKey(
            keyHandle,
            listOf(VerificationRelationshipType.KEY_AGREEMENT, VerificationRelationshipType.AUTHENTICATION)
        )

        val didWeb = DidWeb(ReqwestHttpClient.insecure())

        val documentString = didWeb.generateDidDocument(did, listOf(verificationMethodKey))
        val document = Json.parseToJsonElement(documentString).jsonObject
        val vmm = document["verificationMethod"]?.jsonArray?.get(0)?.jsonObject



        assertEquals("did:web:test.example.com", document["id"]?.jsonPrimitive?.content)
        assertEquals(
            listOf("https://www.w3.org/ns/did/v1", "https://w3id.org/security#EcdsaSecp256r1VerificationKey2019"),
            document["@context"]?.jsonArray?.map { it.jsonPrimitive.content }
        )
        assertEquals(
            listOf("did:web:test.example.com#key-0"),
            document["keyAgreement"]?.jsonArray?.map { it.jsonPrimitive.content })

        assertEquals("did:web:test.example.com", vmm?.get("controller")?.jsonPrimitive?.content)
        assertEquals("did:web:test.example.com#key-0", vmm?.get("id")?.jsonPrimitive?.content)
        assertEquals("EcdsaSecp256r1VerificationKey2019", vmm?.get("type")?.jsonPrimitive?.content)
        assertEquals(true, vmm?.get("publicKeyMultibase")?.jsonPrimitive?.isString)

    }
}


